### ISSUE TITLE
suggestion: support to addons

### ISSUE THREAD

**Issue by zefilipe** [QTodoTxt/QTodoTxt#193]
Hi !

Since Nov 2015, in my new job using windows 7, i started using QTodoTxt !
In the fact its a amazing client gui to todotxt, congratulations loool

But it has a many features that wil be interesting to develop:

-     Statístics by project / contexct
-     Show / hide done tasks in tree of projects and  contexts ?
-     Calculate tasks duration
-     Search done tasks to specific month / date
-     ....

In times ago i developed a many addons to todotxt-client that run on linux shelll, see here
https://github.com/zefilipe/todo.txt-cli/tree/master/addons

It will be interesting if QTodoTxt supports addons, each user use that will be interested  ;-)

There are a way to make addons / develop new features to qtodotxt ?

Thanks !
Regard

**Comment by oroulet** [comment]
probably a good idea. if you send a patch I will have a look.

> how / hide done tasks in tree of projects and contexts ?

I implemented this one through an AllTask filter. It better fits the current architecture. But maybe we could have that as a toggle....., but it does not really fit the current filter stuff

**Comment by kmicc** [comment]
@zefilipe: Hello and welcome to QTodoTxt!
The four features you suggested in the first part of your comment seems to be nice and useful to me. I think they can included in the main codebase if someone implements them. If you want to work on such a particular feature, I suggest creating a new issue for that particular feature, so we can give you some help how things could be implemented.

I took a quick look on your repository. I don't think we can include bash scripts in QTodoTxt due to portability reasons especially regarding towards windows. 

You also wrote something about user defined addons. Maybe this could be achieved  by allowing users to invoke their own user defined (shell) commands directly from QTodoTxt (e.g. via user defined entries in the menu bar)? Seems to be a good idea for me and would allow you to e.g. invoke your bash scripts, run new instances of QTodoTxt on the result or similiar things. Contributions to that are welcome!

**Comment by zefilipe** [comment]
Hi @oroulet  and @kmicc !

Yes, in the fact its a very interesting addons that i develloped to todotxt--client, see in this link a somme screenshots and lite description in use :-)
https://linuxbestchoice.wordpress.com/2012/04/17/todo-txt-o-melhor-gestor-de-tarefas-livre-universal-leve-e-extensivel/

Maybe use google translate page, to get it in english !

Its fine have the AllTask filter ;-)

I know that qtodotxt dont call's linux shell of course :-D

But i can contribute to yours with code ? If yes how to make this ? 

I dont had many free times, but if im interested a little by day, i going develop something interesting !

Regards

**Comment by kmicc** [comment]
Sorry for my answer taking so long. 

Nice to hear that you are interested in developing the one or the other thing here. Of course you can contribute with code :-)

The basic for contributing is a) fork this repository b) implement the things you want to change and upload them to your fork (at best in a separate branch, but that is not mandatory) c) create a pull request (https://github.com/QTodoTxt/QTodoTxt/pulls) on your changes to get them merged into the main repository. If you implemented several different things, please create several pull requests, each containing only one single thing.

If you want to get feedback / help before implementing something, create an issue and tell us what you are going to do.

**Comment by zefilipe** [comment]
Hi @oroulet and @kmicc !

We can continue to talk about this by e-mail ?
I Think that we will exchange knowledge, suggestions ... by it !
And this issue, will leave closed, unless to reopen :-)

My e-mail is:
zefilipe.groups@gmail.com

Regards

**Comment by oroulet** [comment]
Using an issue to discuss a feature is fine otherwise you can use the mailing list. But I am really unsure how to implement a plugin solution. What do you need? A system to write script that are called when clicking on toolbar might be easy.  It should also be possible to call plugins when saving file. Something else?

**Comment by kmicc** [comment]
> A system to write script that are called when clicking on toolbar might be easy.

That's quite close to what I was thinking about: A system to *call* scripts / applications / shell commands / ... written somewhere else when clicking on toolbar / menu item / ... .

**Comment by zefilipe** [comment]
Hi !

I know that an issue to discuss its bad, but i dont found other way to make this :-(

Do you had a mailing list for QTodoTxt ?
If yes, whats subscription e-mail ?

No, sommethink i want discuss a somme tecnical details about the code, issues ... ;-)

Regards

**Comment by oroulet** [comment]
I think an issue is fine too discuss technical things. What do you want to
do in a plugin?
Otherwise look for todotxt in Google groups. The email address is also
written everywhere in code and readme

**Comment by zefilipe** [comment]
Good morning,

Sorry, but in the fact isn't my objective with this subject, to enter spam in this issue ... !

I know and im a subscriber, of todotxt google groups, but in the fact in the recent past, i post one question, but i dont had answer from this team, but by another member, that isn't satisfactory :-(

I do not write most about this subject here, ok ?
I wrote by todotxt google groups, and if get answer ok ;-)

Thanks for all !
Regards

**Comment by oroulet** [comment]
There is something wrong here. I cannot find you in the group and there has
not been any messages since january 15th.

https://groups.google.com/forum/#!forum/qtodotxt

**Comment by zefilipe** [comment]
Yes, are something wrong, i thinked that is todotxt google group, and now i had joined to qtodotxt google group !

Thanks ;-)

**Comment by kmicc** [comment]
The idea of an add-on system for QTT is not wrong, so there is no need to close this issue.

By the way: I'm *not* suscribed to the google group, so there won't be any answers from me there.
